### DISCUSSION TITLE
(middlware/persist): Persisted State Type

### DISCUSSION THREAD

**Discussion by tcconstantin** [pmndrs/zustand#1329] [Q&A] [answered]
Hi, regarding [merge](https://github.com/pmndrs/zustand/blob/main/src/middleware/persist.ts#L65) [migrate](https://github.com/pmndrs/zustand/blob/main/src/middleware/persist.ts#L70), is it an idea to change the type of `persistedState` (currently `unknown`)  with a custom one?

**Comment by dai-shi** [comment, marked as answer]
No, it's really unknown because it's from outside. I think there's a discussion detailed about it.

**Reply by tcconstantin** [reply]
``` ts
type Store = {
  count: number;
};

type Persist = {
  prop: number;
};

export const useStore = create<Store>()(
  persist(
    // ^ persist<Persist> - user will provide this type
    () => ({
      count: 0,
    }),
    {
      name: 'count',
      partialize({ count }) {
        return { prop: count + 1 };
      },
      merge(persistedState) {
        const { prop } = persistedState as Persist;
        // ^ is this the preferred approach?

        return {
          count: prop,
        };
      },
    },
  ),
);
```

**Reply by dai-shi** [reply]
Yeah, in addition, you can create a type guard function and handle runtime error gracefully.

**Reply by Toshinaki** [reply]
@dai-shi 
So is there a recommended way to implement this?
Or any example code in the docs?

**Reply by dai-shi** [reply]
A recommendation is to use a type guard or maybe some libraries like zod.
I don't think we have docs for it. cc: @dbritto-dev

**Reply by dbritto-dev** [reply]
> A recommendation is to use a type guard or maybe some libraries like zod.
> I don't think we have docs for it. cc: @dbritto-dev 

Yeah, let me work on that
